File: src/main/java/com/github/anba/es6draft/runtime/objects/Generator.java
/**
 * Copyright (c) 2012-2013 André Bargull
 * Alle Rechte vorbehalten / All Rights Reserved.  Use is subject to license terms.
 *
 * <https://github.com/anba/es6draft>
 */
package com.github.anba.es6draft.runtime.objects;

import static com.github.anba.es6draft.runtime.internal.Properties.createProperties;

import com.github.anba.es6draft.runtime.Realm;
import com.github.anba.es6draft.runtime.internal.Initialisable;
import com.github.anba.es6draft.runtime.internal.Properties.Attributes;
import com.github.anba.es6draft.runtime.internal.Properties.Prototype;
import com.github.anba.es6draft.runtime.internal.Properties.Value;
import com.github.anba.es6draft.runtime.types.Intrinsics;
import com.github.anba.es6draft.runtime.types.builtins.OrdinaryObject;

/**
 *
 */
public class Generator extends OrdinaryObject implements Initialisable {
    public Generator(Realm realm) {
        super(realm);
    }

    @Override
    public void initialise(Realm realm) {
        createProperties(this, realm, Properties.class);
    }

    public enum Properties {
        ;

        @Prototype
        public static final Intrinsics __proto__ = Intrinsics.FunctionPrototype;

        /**
         * Generator.constructor
         */
        @Value(name = "constructor")
        public static final Intrinsics constructor = Intrinsics.GeneratorFunction;

        /**
         * Generator.prototype
         */
        @Value(name = "prototype", attributes = @Attributes(writable = false, enumerable = false,
                configurable = false))
        public static final Intrinsics prototype = Intrinsics.GeneratorPrototype;
    }
}

File: src/main/scripts/generator.js
/*
 * Copyright (c) 2012-2014 André Bargull
 * Alle Rechte vorbehalten / All Rights Reserved.  Use is subject to license terms.
 *
 * <https://github.com/anba/es6draft>
 */

(function Generator() {
"use strict";

const global = %GlobalObject();

const {
  Object, Symbol, TypeError,
} = global;

const {
  defineProperty: Object_defineProperty,
  getPrototypeOf: Object_getPrototypeOf,
} = Object;

const Generator = Object.getPrototypeOf(function*(){});

Object.defineProperty(Generator, Symbol.hasInstance, {
  value(O) {
    // OrdinaryHasInstance() without steps 1-2
    let C = this;
    if (Object(O) !== O) {
      return false;
    }
    let P = C.prototype;
    if (Object(P) !== P) {
      throw TypeError();
    }
    for (;;) {
      O = Object_getPrototypeOf(O);
      if (O === null) {
        return false;
      }
      if (P === O) {
        return true;
      }
    }
  },
  writable: false, enumerable: false, configurable: true
});

Object.defineProperty(Object.prototype, Symbol.iterator, {
  get() {
    if (typeof this.next === "function") {
      return () => ({__proto__: null, next: () => Object(this.next())})
    }
  },
  set(iter) {
    Object_defineProperty(this, Symbol.iterator, {
      __proto__: null,
      value: iter,
      writable: true, enumerable: true, configurable: true
    });
  },
  enumerable: false, configurable: true
});

})();
